pytorch_lightning/utilities/auto_restart.py

```diff
@@ -447,10 +455,25 @@ def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate:
             "__pl_restart_meta": {"sampler_name0": state_dict0, "sampler_name1": state_dict1},
         }
     """
-    data = default_collate(samples)
-    if not isinstance(dataset, (CaptureIterableDataset, CaptureMapDataset)):
-        return data
-    metadata = dataset.state_dict()
+    data = collate_fn(samples)
+    metadata = None
+    if fault_tolerant_mode.is_automatic:
+        metadata = dataset.state_dict()
+    else:
+        state_dict_fn = getattr(dataset, "state_dict", None)
+        info = get_worker_info()
+        worker_id = info.id if info else 0
+        if state_dict_fn is not None:
+            metadata = state_dict_fn()
+            if worker_id not in metadata:
+                if info and info.num_workers > 1:
+                    raise MisconfigurationException(
+                        f"The state_dict returned by {dataset} needs to be indexed by `worker_id` integer keys."
+                    )
+                metadata = {0: metadata}
+        if metadata is None:
+            metadata = {worker_id: {}}
+
     return {"data": data, AutoRestartBatchKeys.PL_RESTART_META: metadata}
```

Comment:
I know this doesn't come from this PR, but still related: for every batch we also get the corresponding state, right? Have you seen how large they are? Asking because they will also be transferred through shared memory. Does it slow down the process a lot? Probably not if you have some io-heavy stuff in your dataset, but what if returning preloaded data?

Reply:
Yes, with the BoringModel, we get a 30 % slowdown. But with Resnet50 on Cifar10, it is close to 0. @awaelchli performed the benchmark.